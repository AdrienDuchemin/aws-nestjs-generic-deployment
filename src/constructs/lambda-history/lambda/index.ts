import { Context, SQSHandler, SQSEvent } from 'aws-lambda'
import { DynamoDB } from 'aws-sdk'

export const handler: SQSHandler = async (event: SQSEvent, context: Context) => {
    console.log({ event, context })

    if(!process.env.TABLE_NAME) {
        throw new Error('no table name given')
    }

    const client = new DynamoDB.DocumentClient()

    for(const record of event.Records){
        const body = record.body ? JSON.parse(record.body) : { }
        await client.put({
            TableName: process.env.TABLE_NAME,
            Item: {
                id: record.messageId,
                text: body.text,
            }
        }).promise()
    }
}