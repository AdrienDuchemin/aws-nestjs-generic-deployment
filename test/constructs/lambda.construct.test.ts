import { join } from 'path'
import { expect as expectCDK, haveResource, SynthUtils, countResources } from '@aws-cdk/assert'
import { Stack } from '@aws-cdk/core'
import { Lambda, LambdaProps } from '../../src/constructs/lambda.construct'

describe('Lambda', () => {
  let stack: Stack
  const props: LambdaProps = {
    path: join(__dirname, './lambda'),
    handler: "function.handler",
  }

  beforeEach(()=> {
    stack = new Stack()
  })

  it('should create one default Lambda', () => {
    new Lambda(stack, 'Lambda', props)

    expectCDK(stack).to(countResources("AWS::Lambda::Function", 1))
    expectCDK(stack).to(haveResource("AWS::Lambda::Function",{
      Handler: "function.handler",
      Timeout: 30,
      Runtime: "nodejs12.x",
    }))
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot()
  })

  it('should create one Lambda with specified timeout', () => {
    const overridenProps: LambdaProps =  {
      ...props,
      timeout: 15,
    }
    new Lambda(stack, 'Lambda', overridenProps)
    
    expectCDK(stack).to(countResources("AWS::Lambda::Function", 1))
    expectCDK(stack).to(haveResource("AWS::Lambda::Function",{
      Timeout: 15,
    }))
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot()
  })
})