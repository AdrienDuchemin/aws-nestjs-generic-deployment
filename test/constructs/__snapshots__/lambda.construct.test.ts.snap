// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lambda should create one Lambda with specified timeout 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters1ebc9d3ac2033816c4abb63e4afd69d350b4aba8704cc9236b82ea520b74f4b0ArtifactHash4A934180": Object {
      "Description": "Artifact hash for asset \\"1ebc9d3ac2033816c4abb63e4afd69d350b4aba8704cc9236b82ea520b74f4b0\\"",
      "Type": "String",
    },
    "AssetParameters1ebc9d3ac2033816c4abb63e4afd69d350b4aba8704cc9236b82ea520b74f4b0S3Bucket5EA66AEF": Object {
      "Description": "S3 bucket for asset \\"1ebc9d3ac2033816c4abb63e4afd69d350b4aba8704cc9236b82ea520b74f4b0\\"",
      "Type": "String",
    },
    "AssetParameters1ebc9d3ac2033816c4abb63e4afd69d350b4aba8704cc9236b82ea520b74f4b0S3VersionKeyD171B821": Object {
      "Description": "S3 key for asset version \\"1ebc9d3ac2033816c4abb63e4afd69d350b4aba8704cc9236b82ea520b74f4b0\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "lambda8B5974B5": Object {
      "DependsOn": Array [
        "lambdaServiceRole494E4CA6",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters1ebc9d3ac2033816c4abb63e4afd69d350b4aba8704cc9236b82ea520b74f4b0S3Bucket5EA66AEF",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters1ebc9d3ac2033816c4abb63e4afd69d350b4aba8704cc9236b82ea520b74f4b0S3VersionKeyD171B821",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters1ebc9d3ac2033816c4abb63e4afd69d350b4aba8704cc9236b82ea520b74f4b0S3VersionKeyD171B821",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "function.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "lambdaServiceRole494E4CA6",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 15,
      },
      "Type": "AWS::Lambda::Function",
    },
    "lambdaServiceRole494E4CA6": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;

exports[`Lambda should create one default Lambda 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters1ebc9d3ac2033816c4abb63e4afd69d350b4aba8704cc9236b82ea520b74f4b0ArtifactHash4A934180": Object {
      "Description": "Artifact hash for asset \\"1ebc9d3ac2033816c4abb63e4afd69d350b4aba8704cc9236b82ea520b74f4b0\\"",
      "Type": "String",
    },
    "AssetParameters1ebc9d3ac2033816c4abb63e4afd69d350b4aba8704cc9236b82ea520b74f4b0S3Bucket5EA66AEF": Object {
      "Description": "S3 bucket for asset \\"1ebc9d3ac2033816c4abb63e4afd69d350b4aba8704cc9236b82ea520b74f4b0\\"",
      "Type": "String",
    },
    "AssetParameters1ebc9d3ac2033816c4abb63e4afd69d350b4aba8704cc9236b82ea520b74f4b0S3VersionKeyD171B821": Object {
      "Description": "S3 key for asset version \\"1ebc9d3ac2033816c4abb63e4afd69d350b4aba8704cc9236b82ea520b74f4b0\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "lambda8B5974B5": Object {
      "DependsOn": Array [
        "lambdaServiceRole494E4CA6",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters1ebc9d3ac2033816c4abb63e4afd69d350b4aba8704cc9236b82ea520b74f4b0S3Bucket5EA66AEF",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters1ebc9d3ac2033816c4abb63e4afd69d350b4aba8704cc9236b82ea520b74f4b0S3VersionKeyD171B821",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters1ebc9d3ac2033816c4abb63e4afd69d350b4aba8704cc9236b82ea520b74f4b0S3VersionKeyD171B821",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "function.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "lambdaServiceRole494E4CA6",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "lambdaServiceRole494E4CA6": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
